File: src/main/java/boj/part2/Lim10971.java
package boj.part2;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Lim10971 {
	private static int min = Integer.MAX_VALUE;
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		int[][] cost = new int[n][n];
		
		for(int i = 0; i < n; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine());
			
			for(int j = 0; j < n; j++) {
				cost[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		
		boolean[] visited = new boolean[n];
		
		for(int i = 0; i < n; i++) {
			visited[i] = true;
			dfs(cost, visited, i, i, n, 0, 1);
			visited[i] = false;
		}
		
		System.out.println(min);
	}

	private static void dfs(int[][] cost, boolean[] visited, int dest, int depart, int n, int expense, int depth) {
		
		if(depth == n) {
			if(cost[dest][depart] != 0) {
				int totalExpense = expense + cost[dest][depart];
				
				if(totalExpense < min) {
					min = totalExpense;
				}
			}
			
			return;
		}
		
		for(int i = 0; i < n; i++) {
			if(!visited[i] && cost[dest][i] != 0 && expense + cost[dest][i] < min) {
				visited[i] = true;
				dfs(cost, visited, i, depart, n, expense + cost[dest][i], depth + 1);
				visited[i] = false;
			}
		}
	}
}

File: src/main/java/boj/part7/Lim10971.kt
package boj.part7

import java.io.BufferedReader
import java.io.InputStreamReader

private var min = Int.MAX_VALUE

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    val w = Array(n) { br.readLine().split(" ").map { it.toInt() }.toIntArray() }
    val visited = BooleanArray(n)

    for (i in 0 until n) {
        explore(i, i, w, visited, 0, 1, n)
        visited[i] = false
    }

    println(min)
}

private fun explore(start: Int, idx: Int, w: Array<IntArray>, visited: BooleanArray, fee: Int, depth: Int, n: Int) {
    visited[idx] = true
    if (depth == n) {
        if (w[idx][start] != 0 && fee + w[idx][start] < min) {
            min = fee + w[idx][start]
        }
        return
    }

    for (i in 0 until n) {
        if (!visited[i] && w[idx][i] != 0) {
            explore(start, i, w, visited, fee + w[idx][i], depth + 1, n)
            visited[i] = false
        }
    }
}
